### PR TITLE
Write/Fetch posts and display in recent view

### PR DESCRIPTION
**Pull request by judyliu14**
- upper right corner contains "+" button that leads to a new page where user can write posts
- "recent view" has been updated where the navigation bar contains "logout" and "+" button, and the main content consists of a list view of all posts that were made

### CODE REVIEW

File: database/DatabaseExampleSwiftUI/DatabaseExample/Shared/Models/Post.swift [hunk 1]
```diff
@@ -14,32 +14,31 @@
 //  limitations under the License.
 //
 
-import UIKit
+import Foundation
 import Firebase
 
-struct Post {
+struct Post: Identifiable {
+  var id = UUID()
```
Comment (morganchen12) [inline database/DatabaseExampleSwiftUI/DatabaseExample/Shared/Models/Post.swift:21]: Is this correct? In its current state, this implementation of Identifiable will treat two posts with the same author, uid, title, and body as different posts.

File: database/DatabaseExampleSwiftUI/DatabaseExample/Shared/Models/Post.swift [hunk 2]
```diff
@@ -14,32 +14,31 @@
...
 
-  init?(snapshot: DataSnapshot) {
-    guard let dict = snapshot.value as? [String: Any] else { return nil }
+  init?(dict: [String: Any]) {
     guard let uid = dict["uid"] as? String else { return nil }
-    guard let author = dict["author"] as? String else { return nil }
     guard let title = dict["title"] as? String else { return nil }
     guard let body = dict["body"] as? String else { return nil }
     let starCount = dict["starCount"] as? Int ?? 0
+    let author = dict["author"] as? String ?? "default author"
```
Comment (morganchen12) [inline database/DatabaseExampleSwiftUI/DatabaseExample/Shared/Models/Post.swift:41]: This is temporary and should eventually be removed, maybe add a TODO comment?

File: database/DatabaseExampleSwiftUI/DatabaseExample/Shared/Models/UserViewModel.swift [hunk 3]
```diff
@@ -25,6 +25,10 @@ class UserViewModel: ObservableObject {
   @Published var alertMessage = ""
   @Published var isLoading = false
   @Published var alert = false
+  @Published var posts: [Post] = []
+  var ref: DatabaseReference!
+  var postRef: DatabaseReference!
```
Comment (morganchen12) [inline database/DatabaseExampleSwiftUI/DatabaseExample/Shared/Models/UserViewModel.swift:30]: `ref` and `postRef` can be lazy private variables.

```swift
lazy private var ref: DatabaseReference = {
  Database.database().reference()
}()
```

File: database/DatabaseExampleSwiftUI/DatabaseExample/Shared/Models/UserViewModel.swift [hunk 4]
```diff
@@ -94,6 +98,31 @@ class UserViewModel: ObservableObject {
...
+    fetchPosts()
+  }
+
+  func fetchPosts() {
+    posts.removeAll()
+    ref = Database.database().reference()
+    postRef = ref.child("posts")
+    refHandle = postRef.observe(DataEventType.value, with: { snapshot in
+      guard let value = snapshot.value as? [String: [String: Any]] else { return }
+      self.posts = value.compactMap { Post(dict: $1) }
```
Comment (morganchen12) [inline database/DatabaseExampleSwiftUI/DatabaseExample/Shared/Models/UserViewModel.swift:122]: Very elegant transformation!

File: database/DatabaseExampleSwiftUI/DatabaseExample/Shared/Views/ContentView.swift [hunk 5]
```diff
@@ -18,7 +18,7 @@ import SwiftUI
 import Firebase
 
 struct ContentView: View {
-  @AppStorage("isSignedIn") var isSignedIn = false
+  @AppStorage("isSignedIn") var isSignedIn = true
```
Comment (morganchen12) [inline database/DatabaseExampleSwiftUI/DatabaseExample/Shared/Views/ContentView.swift:21]: How is this determined / what's the motivation behind this change?

File: database/DatabaseExampleSwiftUI/DatabaseExample/Shared/Views/NewPostsView.swift [hunk 6]
```diff
@@ -0,0 +1,52 @@
+//
+//  NewPostsView.swift
+//  DatabaseExample
+//
+//  Created by Judy Liu on 7/7/21.
+//
```
Comment (morganchen12) [inline database/DatabaseExampleSwiftUI/DatabaseExample/Shared/Views/NewPostsView.swift:6]: Update copyrights

**Review by morganchen12 — COMMENTED** [review]
Don't forget to run the style script!

File: database/DatabaseExampleSwiftUI/DatabaseExample/Shared/Models/Post.swift (hunk 1 above)
Comment (judyliu14) [reply database/DatabaseExampleSwiftUI/DatabaseExample/Shared/Models/Post.swift:21]: I was initially using another class called PostModel but decided to just use UserViewModel instead. So I think the Identifiable property is no longer needed.

File: database/DatabaseExampleSwiftUI/DatabaseExample/Shared/Models/Post.swift (hunk 2 above)
Comment (judyliu14) [reply database/DatabaseExampleSwiftUI/DatabaseExample/Shared/Models/Post.swift:41]: I will revert this back to the original `guard let` declaration since the author is passed to RTDB when writing the post.

File: database/DatabaseExampleSwiftUI/DatabaseExample/Shared/Views/ContentView.swift (hunk 5 above)
Comment (judyliu14) [reply database/DatabaseExampleSwiftUI/DatabaseExample/Shared/Views/ContentView.swift:21]: The declaration of isSignedIn variable here is not required. The purpose for the variable value is mainly used for testing purposes. Initially when I was implementing for login/logout it is more convenient to be logged out as the initial state, and for making and fetching post it is more convenient to be logged in as the initial state.

Comment (morganchen12) [reply database/DatabaseExampleSwiftUI/DatabaseExample/Shared/Views/ContentView.swift:21]: sgtm

File: database/DatabaseExampleSwiftUI/DatabaseExample/Shared/Models/Post.swift (hunk 1 above)
Comment (morganchen12) [reply database/DatabaseExampleSwiftUI/DatabaseExample/Shared/Models/Post.swift:21]: Having an `id` variable is required, but instead of assigning it to a randomly-generated identifier, you should use the key from Firebase Database. This way, you can construct two instances of the same Post from a database query and they'll be treated as the same value by anything that observes the Identifiable protocol.

File: database/DatabaseExampleSwiftUI/DatabaseExample/Shared/Models/UserViewModel.swift [hunk 7]
```diff
@@ -100,29 +107,40 @@ class UserViewModel: ObservableObject {
   }
 
   func post(title: String, body: String) {
-    ref = Database.database().reference()
+    // check if both title and body are completed
+    if title.isEmpty || body.isEmpty {
+      showAlertMessage(message: "Title or body cannot be empty.")
```
[the remark this answers is not in the document]
Comment (paulb777) [reply database/DatabaseExampleSwiftUI/DatabaseExample/Shared/Models/UserViewModel.swift:112]: And similar with login and signup messages.

File: database/DatabaseExampleSwiftUI/DatabaseExample/DatabaseExample.xcodeproj/project.pbxproj [hunk 8]
```diff
@@ -9,6 +9,8 @@
 /* Begin PBXBuildFile section */
 		C70B63FF2696B29400851A91 /* NewPostsView.swift in Sources */ = {isa = PBXBuildFile; fileRef = C70B63FE2696B29400851A91 /* NewPostsView.swift */; };
 		C70B64002696B29400851A91 /* NewPostsView.swift in Sources */ = {isa = PBXBuildFile; fileRef = C70B63FE2696B29400851A91 /* NewPostsView.swift */; };
+		C799B5F0269CECF800A43C40 /* GoogleService-Info.plist in Resources */ = {isa = PBXBuildFile; fileRef = C799B5EF269CECF800A43C40 /* GoogleService-Info.plist */; };
```
[the remark this answers is not in the document]
Comment (judyliu14) [reply database/DatabaseExampleSwiftUI/DatabaseExample/DatabaseExample.xcodeproj/project.pbxproj]: After removing the GoogleService-Info.plist file the code is still present in the project.pbxproj file. Is there a way I can fully remove this?

[the remark this answers is not in the document]
Comment (morganchen12) [reply database/DatabaseExampleSwiftUI/DatabaseExample/DatabaseExample.xcodeproj/project.pbxproj]: From the project navigator in Xcode, delete the file reference.

[the remark this answers is not in the document]
Comment (judyliu14) [reply database/DatabaseExampleSwiftUI/DatabaseExample/DatabaseExample.xcodeproj/project.pbxproj]: Thanks I believe that worked. The new commits no longer have the GoogleService-Info.plist references in the file.

File: database/DatabaseExampleSwiftUI/DatabaseExample/Shared/Models/UserViewModel.swift [hunk 9]
```diff
@@ -25,17 +25,36 @@ class UserViewModel: ObservableObject {
...
-    alertMessage = "Email or password cannot be empty."
+  private lazy var postRef: DatabaseReference = {
+    Database.database().reference().child("posts")
+  }()
+
+  private lazy var userPostRef: DatabaseReference = {
+    Database.database().reference().child("user-posts").child(getuid())
+  }()
+
+  var refHandle: DatabaseHandle?
```
Comment (morganchen12) [inline database/DatabaseExampleSwiftUI/DatabaseExample/Shared/Models/UserViewModel.swift:42]: `private`?

File: database/DatabaseExampleSwiftUI/DatabaseExample/Shared/Models/Post.swift [hunk 10]
```diff
@@ -14,33 +14,35 @@
...
     self.author = author
     self.title = title
     self.body = body
     starCount = 0
   }
 
-  init?(snapshot: DataSnapshot) {
-    guard let dict = snapshot.value as? [String: Any] else { return nil }
+  init?(dict: [String: Any]) {
+    guard let id = dict["id"] as? String else { return nil }
```
Comment (morganchen12) [inline database/DatabaseExampleSwiftUI/DatabaseExample/Shared/Models/Post.swift:38]: Instead of doing this, you could rewrite the initializer to take two parameters:

```swift
init?(id: String, dict: [String: Any]) { /* ... */ }
```

Then, later on, the `id` doesn't need to be a part of the dictionary, which will save you some storage space in the database.

File: database/DatabaseExampleSwiftUI/DatabaseExample/Shared/Models/UserViewModel.swift [hunk 11]
```diff
@@ -94,6 +114,59 @@ class UserViewModel: ObservableObject {
...
+    }
+
+    // begin loading animation
+    withAnimation {
+      self.isLoading.toggle()
+    }
+
+    let userID = getuid()
+    guard let key = postRef.childByAutoId().key else { return }
+    let post = ["id": key,
```
Comment (morganchen12) [inline database/DatabaseExampleSwiftUI/DatabaseExample/Shared/Models/UserViewModel.swift:132]: If you follow my suggestion above, the first key-value pair (`"id": key`) can be deleted.

File: database/DatabaseExampleSwiftUI/DatabaseExample/Shared/Models/UserViewModel.swift [hunk 12]
```diff
@@ -94,6 +114,59 @@ class UserViewModel: ObservableObject {
...
+
+  func fetchMyPosts() {
+    // read data by listening for value events
+    refHandle = userPostRef.observe(DataEventType.value, with: { snapshot in
+      // retrieved data is of type dictionary of dictionary
+      guard let value = snapshot.value as? [String: [String: Any]] else { return }
+      // sort dictionary by keys (most to least recent)
+      let sortedValues = value.sorted(by: { $0.key > $1.key })
+      // store content of sorted dictionary into "posts" variable
+      self.myPosts = sortedValues.compactMap { Post(dict: $1) }
```
Comment (morganchen12) [inline database/DatabaseExampleSwiftUI/DatabaseExample/Shared/Models/UserViewModel.swift:167]: This line would then become

```swift
self.myPosts = sortedValues.compactMap { Post(id: $0, dict: $1) }
```

File: database/DatabaseExampleSwiftUI/DatabaseExample/Shared/Models/Post.swift (hunk 10 above)
Comment (judyliu14) [reply database/DatabaseExampleSwiftUI/DatabaseExample/Shared/Models/Post.swift:38]: I feel that the "id" information would be helpful to retrieve a specific post and make modifications based on it. Such as incrementing stars and comments. Or is there another way to do so without the id information?

Comment (judyliu14) [reply database/DatabaseExampleSwiftUI/DatabaseExample/Shared/Models/Post.swift:38]: Actually I see that the id will still be included as part of the post object in the last step. I see, thanks for pointing it out!

**Review by morganchen12 — APPROVED** [review]
LGTM!

**Review by paulb777 — APPROVED** [review]
